### ISSUE TITLE
WIDGET_CONTROL - Variable is undefined

### ISSUE THREAD

**Issue by Sandipan-Borthakur** [gnudatalanguage/gdl#1555]
Hii, @GillesDuvert , I am running the BinMag software ([access here](https://www.astro.uu.se/~oleg/binmag.html)) using GDL, and most of the functions work fine, but when I try to input some values in a text box in GUI, an error pops up saying - Variable is undefined. When I checked the binmag6.pro file, the error occurs in a function defined in the following way - 

```
Function GetVal,event_id,i0,i1,valid,nonneg=nonneg

@allcom6

WIDGET_CONTROL,event_id,get_value=sval,/HOURGLAS
sval=sval[0]
valid=1

if strpos(sval,'*') ne -1 then begin
  sval=strmid(sval,0,strpos(sval,'*'))+ $
       strmid(sval,strpos(sval,'*')+1)
  if strlen(sval) gt 7 then sval=string(strtrim(sval,1),form='(A7)')

  WIDGET_CONTROL,event_id,set_value=sval
  val=double(sval)

  if tabact le synmax-1 then begin
    i0=0 & i1=synmax-1
  endif else $
  if tabact ge synmax and tabact le synmax+obsmax-1 then begin
    i0=synmax & i1=synmax+obsmax-1
  endif else $
  if tabact ge synmax+obsmax then begin
    i0=synmax+obsmax & i1=synmax+obsmax+binmax-1
  endif
endif else begin
  i0=tabact & i1=tabact
  val=double(sval)
endelse

if keyword_set(nonneg) and val lt 0 then begin
  s=dialog_message('Negative value not allowed')
  val=-1
  valid=0
endif

return,val
end

```
The error occurs, mentioning that sval is undefined. There are multiple functions within the binmag6.pro file which uses WIDGET_CONTROL, and I think this error will occur for every text box in the GUI if I try to input some values. It seems evident to me that sval is not defined, but this software has been used for many years now. Do you know what the issue is here?

**Comment by GillesDuvert** [comment]
Hi, there was a similar problem just solved in #1535. Do you have a version of GDL that has those patches? More recent, I mean.
If yes, then this is another bug. 
besides, is this on a Windows machine ?

**Comment by Sandipan-Borthakur** [comment]
Hi, thanks for replying. I have version 1.0.2. As you may remember, I raised #1539  regarding the installation of GDL last week. That is when I installed GDL by cloning the git repository. And as I can see, #1535 closed two weeks ago; I believe I have the latest version of GDL. Maybe you can suggest me to check for something specific within GDL to verify if I have the latest version. Also, I am using Ubuntu 20.04.5 LTS.

**Comment by GillesDuvert** [comment]
please remove "/hourglass" at line 3667   `WIDGET_CONTROL,event_id,get_value=sval,/HOURGLAS`  -> `WIDGET_CONTROL,event_id,get_value=sval`

this pending a bug patch of course.

**Comment by Sandipan-Borthakur** [comment]
I removed it, and I do not get an error now. I might have to contact you for a few more minor issues. I apologise for bothering you so much. You can let me know if it is too much I ask from you. For now, I am closing this issue, but I know there are some more issues which I encountered today but do not remember exactly.

**Comment by Sandipan-Borthakur** [comment]
Hii
I ran into another issue while trying to fit the spectrum using the BinMag software. In the GUI, you can access by going to Tools and then Profile fit to test it. If you have binmag.pro file, then the error occurs at line 7287 - 

`WIDGET_CONTROL,lft.base,tbl_get_offset=ofs`

The error is the following - 

`WIDGET_CONTROL: Invalid widget identifier: 999`

**Comment by GillesDuvert** [comment]
Hi, I'll answer you in a sec, just to say that I've tried with IDL and there are many errors too, meaning that the software needs a good dose of updating. For example:
```
% XMANAGER: Caught unexpected error from client application. Message follows...
% Keyword CURRENTCOL not allowed in call to: PICKCOLOR
% Execution halted at:  BINMAG_EVENT     7955 /tmp/BinMag6/binmag6.pro
%                       XMANAGER_EVLOOP_STANDARD  478 /opt/idl/idl84/lib/xmanager.pro
%                       XMANAGER          710 /opt/idl/idl84/lib/xmanager.pro
%                       $MAIN$           8488 /tmp/BinMag6/binmag6.pro

```

This a problem with the color chooser, when the file `coy_binmag/pickcolor.pro` is not the first in the `!PATH` (this procedure has changed since the writing of binmag6 in the IDLAstro repository and its arguments are different now)
But even with this `coy_binmag/pickcolor.pro`  compiled, we get
```
% XMANAGER: Caught unexpected error from client application. Message follows...
% Attempt to call undefined procedure: 'PICKCOLOR_SELECT_COLOR'.
% Execution halted at:  XMANAGER_EVLOOP_REAL_MODAL  506 /opt/idl/idl84/lib/xmanager.pro
%                       XMANAGER          706 /opt/idl/idl84/lib/xmanager.pro
%                       PICKCOLOR         476 /tmp/BinMag6/coy_binmag/pickcolor.pro
%                       BINMAG_EVENT     7955 /tmp/BinMag6/binmag6.pro
%                       XMANAGER_EVLOOP_STANDARD  478 /opt/idl/idl84/lib/xmanager.pro
%                       XMANAGER          710 /opt/idl/idl84/lib/xmanager.pro
%                       $MAIN$           8488 /tmp/BinMag6/binmag6.pro
```

IDL finds also:
```
% XMANAGER: Caught unexpected error from client application. Message follows...
% WIDGET_CONTROL: Illegal keyword value for DRAW_XSIZE.
% Execution halted at:  XREWIN           3978 /tmp/BinMag6/binmag6.pro
%                       BINMAG_EVENT     5747 /tmp/BinMag6/binmag6.pro
%                       XMANAGER_EVLOOP_STANDARD  478 /opt/idl/idl84/lib/xmanager.pro
%                       XMANAGER          710 /opt/idl/idl84/lib/xmanager.pro
%                       $MAIN$           8488 /tmp/BinMag6/binmag6.pro
```
 Note that GDL gives the same errors, which is reassuring.

But anyways using the color picker widget will not work with GDL anyway, as BinMag runs into a severe limitation of GDL caused by our use of the 'plplot' library, that is currently limited to 100 'drawables'. We hope to solve that in the future. So, no color widget.

**Comment by GillesDuvert** [comment]
`WIDGET_CONTROL: Invalid widget identifier: 999` : GDL apparently does not mimic exactly the timing of IDL when widgets are destroyed. This is a nice bug, thanks for pointing to it.
Currently there is no simple way to get around this bug by changing the binmag6 code, although not destroying but just unmapping widgets in procedure `binmg_cleanup` (line 5641)  *and* making them visible again when needed would be feasible.

**Comment by Sandipan-Borthakur** [comment]
Hii @GillesDuvert , thanks for the detailed response. I apologise for the late response, but unfortunately, my older laptop with the GDL stopped working. I have now installed it on my new laptop!!

I didn't know that BinMag causes issues with IDL too. I might have to talk to my supervisor regarding that because I have no idea about IDL, and I will not be able to modify it by myself. I checked the pickcolour issue, but I didn't get the errors you got, probably because of how you added path in the gdlstartup.pro file. But I got a different error with pickcolour - 
![Screenshot from 2023-03-20 14-42-59](https://user-images.githubusercontent.com/92291243/226342157-3389ee77-a1ed-4ed5-9937-014c2bf4be15.png)
This is the content of my gdlstartup.pro file-
`!PATH=!PATH+':'+Expand_Path('+~/Downloads/astron/')+':'+Expand_Path('+~/Downloads/BINMAG/')+':'+Expand_Path('+~/Downloads/SYNTH3/')+':'+Expand_Path('+~/Downloads/sme/') 
`
When I upload multiple observation files and/or synthetic spectrum, the software outputs different colours for each spectrum.

Regarding your second point, I will have to get back to you because, honestly, I am not quite sure I understand it. It is not because of you, but I do not quite understand how the widgets work.

**Comment by GillesDuvert** [comment]
@Sandipan-Borthakur I have access to an idl 8.4 and made some tests around the widgets logic in binmag6, how they appear, how they close, etc, and I came to the conclusion that GDL was not foolproof concerning the handling of 'blocking' applications as described [here](https://www.l3harrisgeospatial.com/docs/XMANAGER.html). In particular, commands stated **after** a call to `xmanager, /Block`  (which is the default) _may not be executed at all._
This is certainly rare as we know there are many 'blocking' widgets applications running OK, but binmag6's is written in such a way that it triggers this unwanted behaviour, that points to a subtle bug in GDL event-processing loop. Solving binmag6 problem will indeed greatly improve GDL's reliability. Until then binmag6 will not work with GDL and I'm sorry for that. However, stay tuned.

The Pickcolor problem is, as I said, due to a limitation of the graphics library (plplot) used by GDL. This is all the more complicated to solve as it is independent from us and the plplot library maintainers are not active. (For the record, It would only suffice to push the value `PL_NSTREAMS` in their `plplotP.h` to something much larger than their default of `100`.  And recompile plplot of course.) 

I had also issues when running binmag6 with IDL, for example lots of widgets are not writeable, but I'm not familiar with the program and have not installed synthmag, so will not comment further.

**Comment by GillesDuvert** [comment]
As you are with a supervisor and probably in hurry, I advice to try to advance in your project using IDL if possible. The GDL behaviour should be made better soon, but I cannot give  a date.

**Comment by Sandipan-Borthakur** [comment]
Hii @GillesDuvert , yes, I think it will be a good idea to use IDL for now. But thank you for dedicating so much of your time to solving my problem. I sincerely apologise for the trouble. I will certainly keep an eye on GDL and use it in the future. I am closing this issue now.

**Comment by GillesDuvert** [comment]
@Sandipan-Borthakur with #1558 **GDL is able to run BinMag6**, at least to a good extent (not tested all)

I found another problem with IDL (same with GDL): sometimes one gets
```
% WIDGET_CONTROL: Illegal keyword value for DRAW_XSIZE.
% Execution halted at: XREWIN           3978 /tmp/BinMag6/binmag6.pro
```
This happens when setting the percentage of `Window [%]` widget to less than 20.
GDL does not have an error handler ready to that, IDL does.

**Comment by Sandipan-Borthakur** [comment]
Dear @GillesDuvert , that's amazing. I didn't expect you will solve this issue this quickly. Yes, I am aware of the issue caused by changing the window %, but I thought that won't be essential for now.  I apologise for asking a naive question, but how can I now add these edits to my version of GDL? Should I use `git pull gnudatalanguage` or clone gnudatalanguage and start from the beginning?

**Comment by GillesDuvert** [comment]
if you cloned gnudatanguage then a git pull shoudl suffice, since PR #1558 is now merged in the master (trunk).

**Comment by Sandipan-Borthakur** [comment]
Hii @GillesDuvert , I used git pull to update the changes in my repo and I was testing BinMag, but I found two errors -  
![image](https://user-images.githubusercontent.com/92291243/226835950-7352318c-4ebe-4401-9da0-252c317a8de4.png)
This happened when I tried profile fit from the tools section. 
![image](https://user-images.githubusercontent.com/92291243/226837945-8d04370e-d026-457c-8d3d-cfb7b0afe639.png)
This error occurred when I tried spectrum synthesis. It uses SYNTH3, so you might not be able to use it. (Just informing you about the error and not pushing you to solve it now). I think most of the errors lie within BINMAG_EVENT, which is connected to XMANAGER, other than that GDL is working perfectly.

**Comment by GillesDuvert** [comment]
hm, this seems yet to be related to a defect in the widget handling by GDL, despite the advances made. So your "case" keeps to be interesting :smile: 
By the way, BinMag appears to be a useful tool for stellar studies... Wished it did not continuously destroy and recreate its widgets (tha's not needed and is a waste of resources IMHO). And, its documentation, if there is only the README, is way too terse.

**Comment by Sandipan-Borthakur** [comment]
There's barely any documentation, and I agree the README file doesn't help much. I wish I knew how to deal with widgets. I should start learning IDL and widget handling. I will try to understand how BinMag deals with widgets and how widgets work in general and then modify the code. But it will take some time. I believe you work on protoplanetary disks, if I am not looking at the wrong profile!! I am also beginning my PhD work on stars hosting protoplanetary disks, and BinMag is a good tool for studying spectrum. There are other similar tools, but I think the advantage of BinMag is its GUI (if you really want to call it an advantage for now).

**Comment by GillesDuvert** [comment]
I'm retired (freshly) but yes, I've more than notions of stellar spectroscopy, but never had the opportunity to use BinMag for, e.g., abundances fitting etc. Apart binmag6.pro, what should I install to get all the binmag widgets working?

**Comment by Sandipan-Borthakur** [comment]
Ohh, wow. Yes, you must be well acquainted with stellar spectroscopy. If you don't mind me asking, what tools do you use for abundance fitting and parameter estimation? Or do you have your own code to do that?
 
You will need the following - 
1. [IDL Astronomy Library](https://idlastro.gsfc.nasa.gov/ftp/) - I think you already know about this.
2. [SYNTH3](https://www.astro.uu.se/~oleg/synth3.html) - You will need synth3.Linux, s3div.Linux, rds3v.pro. s3div and rds3v have the same functionality, but one is an executable file, and another is an IDL program. But I am unsure which one BinMag uses, so I added both for now.
3. [SME](https://www.stsci.edu/~valenti/sme.html) - You can use the 522 version accessible on the website.

You probably already know, but for the sake of completeness, you will also need a GDL_STARTUP file. The content of the gdlstartup.pro file is - 
`!PATH=!PATH+':'+Expand_Path('+~/fullpath/astron/')+':'+Expand_Path('+~/fullpath/BINMAG/')+':'+Expand_Path('+~/fullpath/SYNTH3/')+':'+Expand_Path('+~/fullpath/sme/')`
I stored all the executables and program files downloaded from the Synth3 website in the SYNTH3 folder, and astron is just the IDL Astronomy Library. Let me know if you need some data files to test it.

**Comment by GillesDuvert** [comment]
hello
I cannot be very active on this right now, but I've installed all the above software, and I cannot reproduce your problem. In the same time I was not able to fit a profile in the BinMag6/example file that come with binmag.
So, how did you exactly got the  point where you get 'WIDGET_CONTROL: Invalid widget identifier' ??

**Comment by Sandipan-Borthakur** [comment]
Hii @GillesDuvert 

It's okay if you do not want to indulge in this. I have installed IDL and I am now able to use BinMag. Unfortunately, I am unable to reproduce the first error using Profile Fit. I think that error occurred when I tried spectrum synthesis. I am unable to use Profile Fit too using GDL. I think the error lies somewhere in the SELECT_ZONE function, which allows us to select a wavelength region(which is not working at all). 

The second error is probably not an actual error. I tested now, and it happens when I try to use spectrum synthesis, but since I am unable to select the wavelength region from the plot, I just clicked Done and after that, the second error pops up.

If you want to test it using IDL, then first load a synthetic file and an observed file. In the synthetic file section, click on labels. Then go to profile fit and select profile function and then, click on find lines. Then you will be able to fit the spectrum.
